lib/pbench/test/unit/server/database/test_server_config_db.py

```diff
@@ -0,0 +1,199 @@
+import pytest
+
+from pbench.server.database.models.server_config import (
+    ServerConfig,
+    ServerConfigBadKey,
+    ServerConfigBadValue,
+    ServerConfigDuplicate,
+    ServerConfigMissingKey,
+)
+
+
+class TestServerConfig:
+    def test_bad_key(self, db_session):
+        """Test server config parameter key validation"""
+        with pytest.raises(ServerConfigBadKey) as exc:
+            ServerConfig.create(key="not-a-key", value="no-no")
+        assert str(exc.value) == "Configuration key 'not-a-key' is unknown"
+
+    def test_construct(self, db_session):
+        """Test server config parameter constructor"""
+        config = ServerConfig(key="dataset-lifetime", value="2")
+        config.add()
+        assert config.key == "dataset-lifetime"
+        assert config.value == "2"
+        assert "dataset-lifetime: '2'" == str(config)
+
+    def test_create(self, db_session):
+        """Test server config parameter creation"""
+        config = ServerConfig.create(key="dataset-lifetime", value="2")
+        assert config.key == "dataset-lifetime"
+        assert config.value == "2"
+        assert "dataset-lifetime: '2'" == str(config)
+
+    def test_construct_duplicate(self, db_session):
+        """Test server config parameter constructor"""
+        ServerConfig.create(key="dataset-lifetime", value=1)
+        with pytest.raises(ServerConfigDuplicate) as e:
+            ServerConfig.create(key="dataset-lifetime", value=2)
+        assert str(e).find("dataset-lifetime") != -1
+
+    def test_get(self, db_session):
+        """Test that we can retrieve what we set"""
+        config = ServerConfig.create(key="dataset-lifetime", value="2")
+        result = ServerConfig.get(key="dataset-lifetime")
+        assert config == result
```

Comment:
Yes, well that's the general problem with transitioning to unit tests. Testing the DB layer means mocking all of the implicit SQLAlchemy attribute behavior: e.g., `key` is a SQLAlchemy `Column` object, not a string. Now, we can strip off the `sqlite3` DB by mocking the `db_session` operations, but it's still not a "true" unit test because we're relying on the SQLAlchemy `__declarative_base__` abstractions to provide typed attributes. How far do we peel the onion before giving up and saying either it's a "pure unit test" or it's "more or less close enough to being a pure unit test"; and if we settle for the latter, is the `sqlite3` really so terrible? (Rhetorical question: I never liked it, it's just not trivial to replace.)

I've wrestled with some of those existential questions, but haven't had the fortitude to face them head-on yet. I'm not excited about doing it here, even though we probably *do* need to face this hydra.